File: ios/Plugin/klaviyoPushPlugin.m
#import <Foundation/Foundation.h>
#import <Capacitor/Capacitor.h>

// Define the plugin using the CAP_PLUGIN Macro, and
// each method the plugin supports using the CAP_PLUGIN_METHOD macro.
CAP_PLUGIN(klaviyoPushPlugin, "klaviyoPush",
           CAP_PLUGIN_METHOD(api, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(identify, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(assignToken, CAPPluginReturnPromise);
)

File: ios/Plugin/klaviyoPushPlugin.swift
import Foundation
import Capacitor
import KlaviyoSwift

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(klaviyoPushPlugin)
public class klaviyoPushPlugin: CAPPlugin {
    private let implementation = klaviyoPush()

    @objc func api(_ call: CAPPluginCall) {
        let value = call.getString("apikey") ?? ""
        Klaviyo.setupWithPublicAPIKey(apiKey: value)
        call.resolve()
    }

    @objc func identify(_ call: CAPPluginCall) {
        let value = call.getString("email") ?? ""
        Klaviyo.sharedInstance.setUpUserEmail(userEmail: value)
        call.resolve()
    }

    @objc func assignToken(_ call: CAPPluginCall) {

        func jsonToData(json: Any) -> Data? {
            do {
                return try JSONSerialization.data(withJSONObject: json, options: JSONSerialization.WritingOptions.prettyPrinted)
            } catch let myJSONError {
                print(myJSONError)
            }
            return nil;
        }

        let value = call.getObject("token") ?? [:]
        if let data = jsonToData(json:value) {
            Klaviyo.sharedInstance.addPushDeviceToken(deviceToken: data)
            call.resolve()
        } else {
            call.reject("Error")
        }
    }
}
